File: TheaterSuggestions/Java/SeatsSuggestions/src/main/java/com/baasie/SeatsSuggestions/Row.java
package com.baasie.SeatsSuggestions;
import  com.baasie.SeatsSuggestions.DeepModel.*;
import lombok.EqualsAndHashCode;

import java.util.ArrayList;
import java.util.List;
import java.util.stream.Collectors;

@EqualsAndHashCode
public class Row {
    private final String name;
    private final List<Seat> seats;

    public Row(String name, List<Seat> seats) {
        this.name = name;
        this.seats = seats
                .stream()
                .map(s -> new Seat(
                        s.rowName(),
                        s.number(),
                        s.pricingCategory(),
                        s.seatAvailability()))
                        .collect(Collectors.toList());
    }

    public List<Seat> seats() {
        return seats;
    }

    public Row addSeat(Seat seat) {
        List<Seat> updatedList = seats.stream().map(s -> s.equals(seat) ? seat : s).collect(Collectors.toList());

        return new Row(name, updatedList);
    }

    public SeatingOptionSuggested suggestSeatingOption(SuggestionRequest suggestionRequest) {

        SeatingOptionSuggested seatingOptionSuggested = new SeatingOptionSuggested(suggestionRequest);

        for (Seat seat : offerAdjacentSeatsNearerTheMiddleOfRow(suggestionRequest)) {
            seatingOptionSuggested.addSeat(seat);

            if (seatingOptionSuggested.matchExpectation()) {
                return seatingOptionSuggested;
            }
        }

        return new SeatingOptionNotAvailable(suggestionRequest);
    }
    public List<Seat> offerAdjacentSeatsNearerTheMiddleOfRow(SuggestionRequest suggestionRequest)
    {
        // 1. offer seats from the middle of the row
        List<SeatWithTheDistanceFromTheMiddleOfTheRow> seatWithTheDistanceFromTheMiddleOfTheRows = new OfferingSeatsNearerMiddleOfTheRow(this).offerSeatsNearerTheMiddleOfTheRow(suggestionRequest);

        return seatWithTheDistanceFromTheMiddleOfTheRows.stream().map(SeatWithTheDistanceFromTheMiddleOfTheRow::seat).collect(Collectors.toList());
    }

    private boolean doNotLookForAdjacentSeatsWhenThePartyContainsOnlyOnePerson(SuggestionRequest suggestionRequest) {
        return suggestionRequest.partyRequested() == 1;
    }

    public Row allocate(Seat seat) {
        List<Seat> newVersionOfSeats = new ArrayList<>();

        seats.forEach(currentSeat -> {
            if (currentSeat.sameSeatLocation(seat)) {
                newVersionOfSeats.add(new Seat(seat.rowName(), seat.number(), seat.pricingCategory(),
                        SeatAvailability.Allocated));
            } else {
                newVersionOfSeats.add(currentSeat);
            }
        });

        return new Row(seat.rowName(), newVersionOfSeats);
    }

    public boolean rowSizeIsEven() {

        return seats().size() % 2 == 0;
    }

    public int theMiddleOfRow() {

        return rowSizeIsEven() ? seats().size() / 2 : Math.abs(seats().size() / 2) + 1;
    }

    public boolean isTheMiddleOfRow(Seat seat) {

        int theMiddleOfRow = theMiddleOfRow();

        if (rowSizeIsEven()) {
            if (Math.abs(seat.number() - theMiddleOfRow) == 0 || seat.number() - (theMiddleOfRow + 1) == 0) {
                return true;
            }
        }
        return Math.abs(seat.number() - theMiddleOfRow) == 0;
    }

    public int distanceFromTheMiddleOfRow(Seat seat) {

        if (rowSizeIsEven())
            return seat.number() - theMiddleOfRow() > 0
                    ? Math.abs(seat.number() - theMiddleOfRow())
                    : Math.abs(seat.number() - (theMiddleOfRow() + 1));

        return Math.abs(seat.number() - theMiddleOfRow());
    }
}

File: TheaterSuggestions/CSharp/SeatsSuggestions/Row.cs
using System.Collections.Generic;
using System.Linq;
using Value;
using SeatsSuggestions.DeepModel;

namespace SeatsSuggestions
{
    public class Row : ValueType<Row>
    {
        public string Name { get; init; }
        public List<Seat> Seats { get; init; }

        public Row(string name, List<Seat> seats)
        {
            Name = name;
            Seats = seats;
        }

        public Row AddSeat(Seat seat)
        {
            return new Row(Name, new List<Seat>(Seats) { seat });
        }

        public SeatingOptionSuggested SuggestSeatingOption(SuggestionRequest suggestionRequest)
        {
            var seatingOptionSuggested = new SeatingOptionSuggested(suggestionRequest);

            foreach (var seat in OfferAdjacentSeatsNearerTheMiddleOfRow(suggestionRequest))
            {
                seatingOptionSuggested.AddSeat(seat);

                if (seatingOptionSuggested.MatchExpectation())
                {
                    return seatingOptionSuggested;
                }
            }

            return new SeatingOptionNotAvailable(suggestionRequest);
        }

        public IEnumerable<Seat> OfferAdjacentSeatsNearerTheMiddleOfRow(SuggestionRequest suggestionRequest)
        {
            // 1. offer seats from the middle of the row
            var seatsWithDistanceFromMiddleOfTheRow =
                new OfferingSeatsNearerMiddleOfTheRow(this).OfferSeatsNearerTheMiddleOfTheRow(suggestionRequest).ToList();

            if (DoNotLookForAdjacentSeatsWhenThePartyContainsOnlyOnePerson(suggestionRequest))
            {
                return seatsWithDistanceFromMiddleOfTheRow.Select(s => s.Seat).ToList();
            }

            // 2. based on seats with distance from the middle of row,
                //    we offer the best group (close to the middle) of adjacent seats
                return new OfferingAdjacentSeatsToMembersOfTheSameParty(suggestionRequest).OfferAdjacentSeats(
                seatsWithDistanceFromMiddleOfTheRow);
        }

        private static bool DoNotLookForAdjacentSeatsWhenThePartyContainsOnlyOnePerson(SuggestionRequest suggestionRequest)
        {
            return suggestionRequest.PartyRequested == 1;
        }

        public Row Allocate(Seat seat)
        {
            var newVersionOfSeats = new List<Seat>();

            foreach (var currentSeat in Seats)
            {
                if (currentSeat.SameSeatLocation(seat))
                {
                    newVersionOfSeats.Add(new Seat(seat.RowName, seat.Number, seat.PricingCategory,
                        SeatAvailability.Allocated));
                }
                else
                {
                    newVersionOfSeats.Add(currentSeat);
                }
            }

            return new Row(seat.RowName, newVersionOfSeats);
        }

        protected override IEnumerable<object> GetAllAttributesToBeUsedForEquality()
        {
            return new object[] {Name, new ListByValue<Seat>(Seats)};
        }
    }
}